CRM/Upgrade/Incremental/php/FiveSixtyFive.php

```diff
@@ -40,6 +40,40 @@ public function upgrade_5_65_alpha1($rev): void {
     $this->addTask('Make Group.frontend_title required', 'alterColumn', 'civicrm_group', 'frontend_title', "varchar(255) NOT NULL COMMENT 'Alternative public description of the group.'", TRUE);
 
     $this->addTask('Update ActionSchedule.limit_to column', 'alterColumn', 'civicrm_action_schedule', 'limit_to', "int COMMENT 'Is this the recipient criteria limited to OR in addition to?'");
+    $this->addTask('Remove Batch Create/Edit Activity Types', 'removeFinancialBatchActivityTypes');
+  }
+
+  /**
+   * dev/financial#216 Remove the Create/Edit Batch Activity Types if they were
+   * never used. Otherwise we leave them there for now, and extension could
+   * re-implement the old behaviour.
+   */
+  public static function removeFinancialBatchActivityTypes($ctx): bool {
+    $atypes = [];
+    $atypes[] = CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_Activity', 'activity_type_id', 'Create Batch');
+    $atypes[] = CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_Activity', 'activity_type_id', 'Edit Batch');
+
+    if (empty(array_filter($atypes))) {
+      return TRUE;
+    }
+
+    // Check for existing Create Batch activities (no need to check for Edit)
+    $exists = CRM_Core_DAO::singleValueQuery('SELECT id FROM civicrm_activity WHERE activity_type_id = %1 LIMIT 1', [
```

Comment:
I put merge-ready only because I'm not sure about only needing to check for Create. I can manually delete a create activity in the UI (or otherwise) and then be left with only batch edit activities, which then after upgrade leads to a site with activities that have no defined activity type. While unlikely, if users CAN do something, they WILL do it.

Also this will need regen but it's not a blocker since I think it's already out of date from elsewhere and needing doing.

Reply:
> While unlikely, if users CAN do something, they WILL do it. 

Fair point, but the consequences are really minor (a lingering Activity Type), and I would rather not slow down too much the upgrade process on large installs.

I can do a regen separately once the PR is merged? They tend to go stale quickly.

Thanks for the review! I know this is by far not the biggest priority in CiviCRM, but I was in hulk-cleanup mode :grimacing:

Reply:
We need more cleanup hulks! 💪🏻